core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -908,6 +981,19 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
                                 }
                             }
                         }
+                    } else if ("primary_terms".equals(currentFieldName)) {
+                        IntArrayList list = new IntArrayList();
+                        while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                            if (token == XContentParser.Token.VALUE_NUMBER) {
+                                list.add(parser.intValue());
+                            } else {
+                                throw new IllegalStateException("found a non-numeric value under [primary_terms]");
+                            }
+                        }
+                        for (int i = 0; i < list.size(); i++) {
+                            builder.primaryTerm(i, list.get(i));
+                        }
```

Comment:
Is the `list` completely unnecessary if we just keep track of an index in the `while` loop and repeatedly call `builder#primaryTerm` as in the `for` loop? That will remove an allocation (and then the JVM does not even need to do escape analysis).